### PR TITLE
making the 3d-padded models more efficient in pytorch

### PR DESCRIPTION
**Pull request by jpata**
- The conversion of 3D-padded shapes `[batch, elems_per_event, features]` with a padding mask to pytorch geometric sparse `[all_elems, features]` was slow. Now this is not done any more when it's not needed
  - The geometric-style models (e.g. gravnet) use sparse data throughout, while the 3D-shape models (e.g. Attention, GNNLSH) use 3D-padded data throughout.
- use pin_memory in the DataLoader, which removes the host-to-device blocking call when doing `batch.to(...)`
- move the validation loop outside the train loop
  - possibly some follow-up with `world_size` is still required

After all changes, here's how the profiles look like on a single batch.

Gravnet, approx. 90ms / batch, including the backward pass:
image

GNNLSH, approx. 125ms / batch, including the backward pass:
image

Longer test on RTX2060S:
```
python3 mlpf/pyg_pipeline.py --config parameters/pyg-cms-small.yaml --dataset cms --gpus 0 --data-dir ~/tensorflow_datasets/ --train --conv-type gnn_lsh --num-epochs 5 --gpu-batch-multiplier 1 --num-workers 1 --prefetch-factor 10

before:
  2%|██▋            | 2656/160000 [06:16<6:11:37,  7.06it/s]

after:
  2%|██▊            | 2776/160000 [05:08<4:50:52,  9.01it/s]
```

### CODE REVIEW

**Comment by jpata** [comment]
After pinning the memory in the data loader, on 1xRTX2060S:
```
$ python3 mlpf/pyg_pipeline.py --config parameters/pyg-cms-small.yaml --dataset cms --gpus 0 --data-dir ~/tensorflow_datasets/ --train --conv-type gnn_lsh --num-epochs 10 --gpu-batch-multiplier 1 --num-workers 1 --prefetch-factor 10 --ntrain 100 --nvalid 100

GNNLSH:
	INFO:mlpf:Rank 0: epoch=10 / 10 train_loss=31.0352 valid_loss=34.5560 stale=0 time=0.44m eta=0.0m
	INFO:mlpf:Done with training. Total training time on device 0 is 4.491min

$ python3 mlpf/pyg_pipeline.py --config parameters/pyg-cms-small.yaml --dataset cms --gpus 0 --data-dir ~/tensorflow_datasets/ --train --conv-type gravnet --num-epochs 10 --gpu-batch-multiplier 1 --num-workers 1 --prefetch-factor 10 --ntrain 100 --nvalid 100

Gravnet:
	INFO:mlpf:Rank 0: epoch=10 / 10 train_loss=29.3231 valid_loss=33.1442 stale=0 time=0.38m eta=0.0m
	INFO:mlpf:Done with training. Total training time on device 0 is 3.884min
```

**Comment by jpata** [comment]
```
$ python3 mlpf/pyg_pipeline.py --config parameters/pyg-cms-small.yaml --dataset cms --gpus 0 --data-dir ~/tensorflow_datasets/ --train --conv-type gnn_lsh --num-epochs 5 --gpu-batch-multiplier 1 --num-workers 1 --prefetch-factor 10 --ntrain 100 --nvalid 100

after PR:
	INFO:mlpf:Rank 0: epoch=5 / 5 train_loss=37.0878 valid_loss=40.5538 stale=0 time=0.44m eta=0.0m
	INFO:mlpf:Done with training. Total training time on device 0 is 2.261min

before PR:
	INFO:mlpf:Rank 0: epoch=5 / 5 train_loss=36.7364 valid_loss=40.1270 stale=0 time=0.49m eta=0.0m
	INFO:mlpf:Done with training. Total training time on device 0 is 2.59min
```
